ssl/statem/extensions_srvr.c

```diff
@@ -127,7 +127,7 @@ int tls_parse_ctos_server_name(SSL *s, PACKET *pkt, unsigned int context,
         return 0;
     }
 
-    if (!s->hit) {
+    if (!s->hit || SSL_IS_TLS13(s)) {
```

Comment:
Why are we doing this if we have TLSv1.3?

Reply:
TLS 1.3 changes (well, or not, depending on who you ask) the semantics of SNI from being bound to the session to being bound to the connection.  Semi-relatedly, we generate new SSL_SESSION objects for TLS 1.3 connections due to some other place we had to modify the structure (I forget the details offhand), so it's always safe to store the new SNI value in the per-connection SSL_SESSION object (which we do in final_server_name()).

Reply:
But how does this concretely change the behaviour? I can't work out how a server will behave different in TLSv1.3 vs TLSv1.2 after this change?

Reply:
For TLS 1.3 resumptions where the client offers a new SNI, SSL_get_servername() will report the correct (new, per-connection) SNI value for TLS 1.3 connections.  If we did not have this chunk, that function would incorrectly report the old SNI value from the previous session/connection.  (Right?)

Reply:
Ah. Right - thanks for the clarification. Where is this change of behaviour between TLSv1.2/TLSv1.3 documeted in the spec?

Reply:
Section 4.2.11 "Pre-Shared Key Extension" has:

```
   In TLS versions prior to TLS 1.3, the Server Name Identification
   (SNI) value was intended to be associated with the session (Section 3
   of [RFC6066]), with the server being required to enforce that the SNI
   value associated with the session matches the one specified in the
   resumption handshake.  However, in reality the implementations were
   not consistent on which of two supplied SNI values they would use,
   leading to the consistency requirement being de facto enforced by the
   clients.  In TLS 1.3, the SNI value is always explicitly specified in
   the resumption handshake, and there is no need for the server to
   associate an SNI value with the ticket.  Clients, however, SHOULD
   store the SNI with the PSK to fulfill the requirements of
   Section 4.6.1.
```

There's also some text in NewSessionTicket about how resuming with a different SNI is a little dicey for now, but there's a draft in the works to say more about how to do it.